File: backend/app/schemas.py
from pydantic import BaseModel
from typing import Optional, List


class Token(BaseModel):
    access_token: str
    token_type: str = "bearer"


class UserCreate(BaseModel):
    username: str
    password: str


class UserOut(BaseModel):
    id: int
    username: str
    role: str

    class Config:
        from_attributes = True


class ModelCreate(BaseModel):
    name: str
    slug: str
    description: Optional[str] = ""
    tags: Optional[str] = ""


# class ModelInfoTag(BaseModel):
#     capabilities: str
#     action: Optional[str] = None
#     target_object: Optional[str] = None
#     language: Optional[str] = None
#     host_device: Optional[str] = None
#     scenario: Optional[str] = None


class ModelInfoMetric(BaseModel):
    parameters: Optional[int] = None
    quantatization: Optional[str] = None
    gpu_memory_gb: Optional[float] = None
    storage_gb: Optional[float] = None


class ModelInfo(BaseModel):
    id: Optional[int] = None
    name: str
    fullname: str
    architecture: str
    framework: str
    version: str
    description: str = ""
    tags: dict
    metrics: ModelInfoMetric


class ModelOut(BaseModel):
    id: int
    name: str
    slug: str
    owner_id: int
    description: str
    tags: str
    download_count: int

    class Config:
        from_attributes = True


class VersionCreate(BaseModel):
    version: str


class VersionOut(BaseModel):
    id: int
    model_id: int
    version: str
    status: str
    is_latest: bool

    class Config:
        from_attributes = True


class FileOut(BaseModel):
    id: int
    version_id: int
    filename: str
    size: int
    sha256: str
    mime: str

    class Config:
        from_attributes = True

File: web/src/schemas.ts
interface ModelTag {
  capabilities: string;
  action?: string;
  target_object?: string;
  language?: string;
  host_device?: string;
  scenario?: string;
}

interface ModelMetric {
  parameters?: number;
  quantatization?: string; // Assuming this is a type of quantization. Adjust accordingly.
  gpu_mempry_gb?: number; // Assuming gpu_mempry_gb is of type 'number'. Adjust as needed.
}

interface ModelInfo {
  architecture: string;
  framework: string;
  version: string;
  tags: { [key: string]: ModelTag }; // This will be a dictionary/map with string keys and values of type `ModelTag`.
  metrics: ModelMetric;
}

interface LegacyModelOut {
  id: number;
  name: string;
  slug: string;
  owner_id: number;
  description: string;
  tags: string;
  download_count: number;
}

export { ModelMetric, ModelInfo, ModelTag, LegacyModelOut };
